R/get_resource.R

```diff
@@ -13,108 +15,70 @@
-opendata_ua <- function() {
-  httr::user_agent("https://github.com/Public-Health-Scotland/phsmethods")
-}
-
-
-#' Check if a resource ID is valid
-#'
-#' @description
-#' Used to attempt to validate a res_id before submitting it to the API
+#' @examples
+#' res_id <- "ca3f8e44-9a84-43d6-819c-a880b23bd278"
+#' filters <- list("HB" = "S08000030", "Month" = "202109")
+#' wanted_cols <- c("HB", "Month", "TotalPatientsSeen")
 #'
-#' @param res_id a resource ID
-#'
-#' @return TRUE / FALSE indicating the validity of the res_id
-check_res_id <- function(res_id) {
-  res_id_regex <-
-    "^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
-
-
-  if (!inherits(res_id, "character")) {
-    return(FALSE)
-  } else if (!grepl(res_id_regex, res_id)) {
-    return(FALSE)
-  } else {
-    return(TRUE)
-  }
-}
-
-#' Creates the URL for the datastore search end-point
-#'
-#' @return a url
-ds_search_url <- function() {
-  httr::modify_url("https://www.opendata.nhs.scot",
-    path = "/api/3/action/datastore_search"
+#' df <- get_resource(res_id = res_id, row_filters = filters, col_select = wanted_cols)
+get_resource <- function(res_id, rows = NULL, row_filters = NULL, col_select = NULL) {
+
+  # check res_id
+  check_res_id(res_id)
+
+  # define query
+  query <- list(
+    id = res_id,
+    limit = rows,
+    q = parse_row_filters(row_filters),
+    fields = parse_col_select(col_select)
   )
-}
 
-#' Creates the URL for the datastore dump end-point
-#'
-#' @param res_id a resource ID
-#' @return a url
-ds_dump_url <- function(res_id) {
-  httr::modify_url("https://www.opendata.nhs.scot",
-    path = glue::glue("/datastore/dump/{res_id}?bom=true")
+  # if dump should be used, use it
+  if (use_dump_check(query, rows))
+    return(dump_download(res_id))
+
+  # if there is no row limit set
+  # set limit to CKAN max
+  if (is.null(query$limit)) query$limit <- 99999
+
+  # remove null values from query
+  null_q_field <- sapply(query, is.null)
+  query[null_q_field] <- NULL
+
+  # fetch the data
+  q <- paste0(paste0(names(query), "=", query), collapse = "&")
+  res_content <- phs_GET("datastore_search", q)
+
+  # if the total number of rows is greater than the
+  # number of rows fetched
+  # AND the user was not aware of this limit (`rows` defaulted to NULL)
+  # warn the user about this limit.
+  total_rows <- res_content$result$total
+  if (is.null(rows) && query$limit < total_rows)
+    cli::cli_warn(c(
+      "Returning the first {query$limit}
+      results (rows) of your query.
+      {total_rows} rows match your query in total.",
+      i = "To get ALL matching rows you will need to download
+      the whole resource and apply filters/selections locally."
+    ))
+
+  # if more rows were requested than received
+  # let the user know
+  if (!is.null(rows) && query$limit > total_rows)
+    cli::cli_warn(c(
+      "You set {.var rows} to {query$limit} but
+      only {total_rows} rows matched your query."
+    ))
```

Comment:
Does this need to be a warning? `cli_alert_warning()` is nicely formatted like a warning but is actually just a message.
```suggestion
    cli::cli_alert_warning(c(
      "You set {.var rows} to {query$limit} but
      only {total_rows} rows matched your query."
    ))
```